### PR TITLE
[MRG] fix joblib dump for numpy arrays of dtype object

### LINKED ISSUE #220: joblib.dump bug with too large object-type arrays
issue from
https://github.com/scikit-learn/scikit-learn/issues/4889
The code
```python
import joblib
import os
import numpy as np
from gzip import GzipFile
from io import BytesIO
from urllib2 import urlopen
from os.path import join
from sklearn.datasets import get_data_home

URL10 = ('http://archive.ics.uci.edu/ml/'
         'machine-learning-databases/kddcup99-mld/kddcup.data_10_percent.gz')

data_home = get_data_home()
kddcup_dir = join(data_home, "test")
samples_path = join(kddcup_dir, "samples")
os.makedirs(kddcup_dir)
f = BytesIO(urlopen(URL10).read())
file = GzipFile(fileobj=f, mode='r')

X = []
for line in file.readlines():
    X.append(line.replace('\n', '').split(','))

file.close()
X = np.asarray(X, dtype=object)
joblib.dump(X, samples_path, compress=9)
X = joblib.load(samples_path)
```
More precisely, it works if X has less than 300000 lines or if X is not of dtype `object`:
```python
Y = X[:300000,:]  ### works
joblib.dump(Y, samples_path, compress=9)
Y = joblib.load(samples_path)
###
Y = X[:400000,:].astype(str)  ### works
joblib.dump(Y, samples_path, compress=9)
Y = joblib.load(samples_path)
###
Y = X[:400000,:]  ### doesnt work
joblib.dump(Y, samples_path, compress=9)
Y = joblib.load(samples_path)
```

The error raised by joblib.load is 
```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-32-25fe17420f08> in <module>()
----> 1 Y = joblib.load(samples_path)

/usr/lib/python2.7/dist-packages/joblib/numpy_pickle.pyc in load(filename, mmap_mode)
    422 
    423         try:
--> 424             obj = unpickler.load()
    425         finally:
    426             if hasattr(unpickler, 'file_handle'):

/usr/lib/python2.7/pickle.pyc in load(self)
    856             while 1:
    857                 key = read(1)
--> 858                 dispatch[key](self)
    859         except _Stop, stopinst:
    860             return stopinst.value

/usr/lib/python2.7/dist-packages/joblib/numpy_pickle.pyc in load_build(self)
    288                         "but numpy didn't import correctly")
    289             nd_array_wrapper = self.stack.pop()
--> 290             array = nd_array_wrapper.read(self)
    291             self.stack.append(array)
    292 

/usr/lib/python2.7/dist-packages/joblib/numpy_pickle.pyc in read(self, unpickler)
    158         array = unpickler.np.core.multiarray._reconstruct(*self.init_args)
    159         with open(filename, 'rb') as f:
--> 160             data = read_zfile(f)
    161         state = self.state + (data,)
    162         array.__setstate__(state)

/usr/lib/python2.7/dist-packages/joblib/numpy_pickle.pyc in read_zfile(file_handle)
     69     assert len(data) == length, (
     70         "Incorrect data length while decompressing %s."
---> 71         "The file could be corrupted." % file_handle)
     72     return data
     73 

AssertionError: Incorrect data length while decompressing <open file '/home/nicolas/scikit_learn_data/test/samples_01.npy.z', mode 'rb' at 0x7efcea714db0>.The file could be corrupted.
```
@lesteve what do you think?

### PR DESCRIPTION
**Pull request by lesteve**
For object dtype, we do not try to save the numpy array in an auxiliary file but just do a pickle.dump in the master pickle file.

Added a numpy array with dtype object in test_compressed_pickle_dump_and_load. Also test the code branch where the array size is bigger than the 'cache_size' argument of joblib.dump.

Fix #220.

### CODE REVIEW

File: joblib/numpy_pickle.py [hunk 1]
```diff
@@ -267,19 +267,14 @@ def save(self, obj):
...
-            try:
+
+            if not obj.dtype.hasobject:
+                self._npy_counter += 1
                 filename = '%s_%02i.npy' % (self._filename,
                                             self._npy_counter)
                 # This converts the array in a container
                 obj, filename = self._write_array(obj, filename)
                 self._filenames.append(filename)
-            except:
```
Comment (lesteve) [inline joblib/numpy_pickle.py]: @GaelVaroquaux do you remember why you wrote this try/except ? What kind of errors is it supposed to tackle ?

Comment (GaelVaroquaux) [inline joblib/numpy_pickle.py]: I don't remember.

However, I do know something: the policy of joblib is to try to never fail. Our philosophy is that adding joblib to your code should not making more fragile. That's why try / excepts are sprinkled in the codebase. The idea is that we might not think about all the possible errors, so if we can't persist the fast way, we persist the standard way.

I am afraid that joblib is becoming more brittle as it is becoming more sophisticated.

**Comment by lesteve** [comment]
> Maybe I'm wrong, but why not adding here a condition "not obj.dtype.hasobject" and keep the previous behavior (using the except) ?

We could do that I am just slightly worried that this try except is not a very clean way of hiding failures. A few weird behaviours we recently saw:
* for Python 2.6 and @aabadie's PR with `with gzip.GZipFile as f`, you get a failure when you exit the context manager, meaning that you actually write a valid .z auxiliary file but then still pickle the array inside the master file because you encountered an error. This auxiliary file is never deleted.
* for Python 3, when trying to dump an array with dtype object, you encountered an error quite early on but you still write an invalid .z auxiliary file with just the header. This auxiliary file is never deleted.
* for Python 2.7, when trying to dump an array with dtype object, you don't even encounter an error and you end up writing str(array). You only see the problem at load time.

@aabadie as an aside, just a gentle reminder: always better to comment in the "Files changed" tab rather than in individual commits.

**Comment by lesteve** [comment]
@GaelVaroquaux I reinstated the try/except for safety's sake. I only catch Exception though, as I seem to remember you very seldom want to catch other exceptions. IIRC KeyboardInterrupt doesn't derive from Exception but do we care here?

**Comment by GaelVaroquaux** [comment]
> I only catch Exception though, as I seem to remember you very seldom
> want to catch other exceptions. IIRC KeyboardInterrupt doesn't derive
> from Exception but do we care here?

Yes, that's good.

**Comment by lesteve** [comment]
All right, merging then.

**Pull request merged**